programs/storage_api/src/storage_contract.rs

```diff
@@ -69,6 +71,44 @@ impl<'a> StorageAccount<'a> {
         Self { account }
     }
 
+    pub fn initialize_mining_pool(&mut self) -> Result<(), InstructionError> {
+        let storage_contract = &mut self.account.state()?;
+        if let StorageContract::Uninitialized = storage_contract {
+            *storage_contract = StorageContract::MiningPool;
+            self.account.set_state(storage_contract)
+        } else {
+            Err(InstructionError::AccountAlreadyInitialized)?
+        }
+    }
+
+    pub fn initialize_replicator_storage(&mut self) -> Result<(), InstructionError> {
```

Comment:
Nit: (especially given the readability benefits) this causes two writes to the account state the first time.

Reply:
Nvm I didn't see the separate init instructions. Somehow it doesn't seem as elegant though.